app/src/main/java/com/asfoundation/wallet/ui/wallets/WalletsInteract.kt

```diff
@@ -0,0 +1,60 @@
+package com.asfoundation.wallet.ui.wallets
+
+import com.asfoundation.wallet.entity.Wallet
+import com.asfoundation.wallet.interact.CreateWalletInteract
+import com.asfoundation.wallet.interact.FetchWalletsInteract
+import com.asfoundation.wallet.repository.SharedPreferencesRepository
+import com.asfoundation.wallet.ui.balance.BalanceInteract
+import com.asfoundation.wallet.ui.iab.FiatValue
+import io.reactivex.Completable
+import io.reactivex.Observable
+import io.reactivex.Single
+import java.math.BigDecimal
+
+class WalletsInteract(private val balanceInteract: BalanceInteract,
+                      private val fetchWalletsInteract: FetchWalletsInteract,
+                      private val createWalletInteract: CreateWalletInteract,
+                      private val preferencesRepository: SharedPreferencesRepository) {
+
+  fun retrieveWalletsModel(): Single<WalletsModel> {
+    val walletsBalance = ArrayList<WalletBalance>()
+    val currentWalletAddress = preferencesRepository.getCurrentWalletAddress()
+    return retrieveWallets().filter { it.isNotEmpty() }
+        .flatMapCompletable { list ->
+          Observable.fromIterable(list)
+              .flatMapCompletable { wallet ->
+                balanceInteract.getTotalBalance(wallet.address)
+                    .take(1)
+                    .flatMapCompletable { fiatValue ->
+                      Completable.fromAction {
+                        walletsBalance.add(WalletBalance(wallet.address, fiatValue,
+                            currentWalletAddress == wallet.address))
+                      }
+                    }
+              }
+        }
+        .toSingle {
+          WalletsModel(getTotalBalance(walletsBalance), walletsBalance.size, walletsBalance)
+        }
+  }
+
+  fun createWallet(): Completable {
+    return createWalletInteract.create()
+        .flatMapCompletable {
+          createWalletInteract.setDefaultWallet(it.address)
+        }
+  }
+
+  private fun getTotalBalance(walletBalance: MutableList<WalletBalance>): FiatValue {
+    var totalBalance = BigDecimal.ZERO
+    for (wallet in walletBalance) totalBalance = totalBalance.add(wallet.balance.amount)
```

Comment:
We could improve this by creating an extension function:

inline fun <T> Iterable<T>.sumByBigDecimal(selector: (T) -> BigDecimal): BigDecimal {
    var sum: BigDecimal = BigDecimal.ZERO
    for (element in this) {
        sum += selector(element)
    }
    return sum
}

and use walletBalance.sumByBigDecimal { it.balance.amount }

Reply:
Don't know if this is the best solution, seems too specific for this case

Reply:
if you don't think a an extenstion would be used in the future at least use map and reduce to simplify readability.